### ISSUE TITLE
Re-design zarr checksum to not rely on "full path"s (and become a proper tree checksum)

### ISSUE THREAD

**Issue by yarikoptic** [dandi/dandischema#119]
Realization (for me) came in https://github.com/dandi/dandi-cli/pull/923#discussion_r814257956 that for zarr checksum a "full" path from the top of zarr asset is used for computing that file/path contribution to the checksum.
Examples of that effect are in https://github.com/dandi/dandischema/blob/master/dandischema/digests/tests/test_zarr.py showing such "full paths" used in ZarrChecksum .  And looking at the code https://github.com/dandi/dandischema/blob/master/dandischema/digests/zarr.py (because [design file I found](https://github.com/dandi/dandi-archive/blob/master/doc/archive/zarr-support-2.md%20) primarily has only TODOs on those aspects) it seems that zarr checksumming relies on a full list of paths and doesn't rely on hierarchical (nested) structure of a checksum of higher directory to be aggregation of checksums of immediate subdirectories + files.

IMHO such design is "suboptimal"
- conceptually in that the checksum of a subdirectory (within a zarr directory) for some reason depends not just on its content but on where in the zarr it resides
-  in "implementation" , as my confusion/question in https://github.com/dandi/dandi-cli/pull/923#discussion_r814257956 showed, because requires knowing the top directory from which to "compute checksum of the subcomponent", so more complicated API.
- "memory demand" -- full relative paths should be stored, thus replicating leading prefixes over and over again for no real reason. 

IMHO ideally underlying checksum datastructure should be a tree, with only immediate children names contributing to checksum computation.

Attn @dchiquito @satra

**Comment by satra** [comment]
@yarikoptic - as i understand the code, it's a generic checksum system at a given level of a tree. it doesn't walk anything or even digest a file. that would be a utility function written on top of those classes. 

to use this code locally, it would have to be combined with a walk and a digest of a file. it would be useful to add such a utility function as well.

and as far as i know, @dchiquito implementation on the api side is a tree. i.e. a checksum at a level of tree is a digest of the structure underneath it.

**Comment by yarikoptic** [comment]
@satra 
Yes, underneath it, but with full paths from top of the tree (at least according to @jwodder ), is that right @dchiquito ?
